client/go/outline/connectivity/connectivity.go

```diff
@@ -94,22 +106,90 @@ func CheckUDPConnectivityWithDNS(client transport.PacketListener, resolverAddr n
 	}
 }
 
+func getDNSRequest() []byte {
+	return []byte{
+		0, 0, // [0-1]   query ID
+		1, 0, // [2-3]   flags; byte[2] = 1 for recursion desired (RD).
+		0, 1, // [4-5]   QDCOUNT (number of queries)
+		0, 0, // [6-7]   ANCOUNT (number of answers)
+		0, 0, // [8-9]   NSCOUNT (number of name server records)
+		0, 0, // [10-11] ARCOUNT (number of additional records)
+		3, 'c', 'o', 'm',
+		0,    // null terminator of FQDN (root TLD)
+		0, 1, // QTYPE, set to A
+		0, 1, // QCLASS, set to 1 = IN (Internet)
+	}
+}
+
 // CheckTCPConnectivityWithHTTP determines whether the proxy is reachable over TCP and validates the
 // client's authentication credentials by performing an HTTP HEAD request to `targetURL`, which must
 // be of the form: http://[host](:[port])(/[path]).
 //
 // Returns nil on success, error on connectivity failure.
-func CheckTCPConnectivityWithHTTP(dialer transport.StreamDialer, targetURL string) error {
-	deadline := time.Now().Add(tcpTimeout)
+func CheckTCPConnectivityWithHTTP(timeout time.Duration, dialer transport.StreamDialer, urlList []string) error {
+	if len(urlList) == 0 {
+		return errors.New("test url list is empty")
+	}
+
+	deadline := time.Now().Add(timeout)
 	ctx, cancel := context.WithDeadline(context.Background(), deadline)
 	defer cancel()
+
+	errCh := make(chan error)
+	for i, targetURL := range urlList {
+		go func() {
+			// We pre-define the start time of the probes to make their timing independent of each other
+			// and to randomize their order. This mitigates some fingerprinting attacks.
+			// The first URL is always started immediately, and given at least 500ms of a head start.
+			// The intent is to not trigger the fallback unless needed.
+			// The other tests start at a random times, uniformly distributed within [500ms, timeout - 500ms]
+			// (we want the last test to have at least 500ms to complete).
+			if i > 0 {
```

Comment:
Shall we pass `i` and `targetURL` as an arguments of `go func` (e.g., `go func(i int, url string) {...}(i, targetURL)`) instead of using them as closures? I'm worried that `i` will always be `len(urlList) - 1` in the function.

A simple verification can be done by making three of URLs to be inaccessible, and see whether it still works.